hazelcast/src/main/java/com/hazelcast/partition/impl/ForceUpdateSyncWaitingReplicaVersionsOperation.java

```diff
@@ -0,0 +1,128 @@
+        final int partitionId = getPartitionId();
+        final InternalPartitionService partitionService = getService();
+        // returns the internal array itself, not the copy
+        final long[] versions = partitionService.getPartitionReplicaVersions(partitionId);
+
+        final ILogger logger = getLogger();
+
+        if (reason == PartitionReplicaChangeReason.MEMBER_REMOVED) {
+            final int lostReplicaIndex = getLostReplicaIndex(versions);
+            if (lostReplicaIndex > 0) {
+                if (logger.isFinestEnabled()) {
+                    logger.finest(
+                            "Partition replica is lost! partitionId=" + partitionId + " lostReplicaIndex=" + lostReplicaIndex
+                                    + " replicaVersions=" + Arrays.toString(versions));
+                }
+
+                overwriteLostReplicaVersionsWithFirstAvailableVersion(versions, lostReplicaIndex);
+            }
+
+            sendPartitionLostEvent(partitionId, lostReplicaIndex);
+        } else {
+            if (logger.isFinestEnabled()) {
+                logger.finest("Resetting all SYNC_WAITING Versions. partitionId=" + partitionId + " versionsBeforeReset="
+                        + Arrays.toString(versions));
+            }
+
+            resetSyncWaitingVersions(versions);
+        }
+    }
+
+    private int getLostReplicaIndex(final long[] versions) {
+        int biggestLostReplicaIndex = 0;
+
+        for (int replicaIndex = 1; replicaIndex <= versions.length; replicaIndex++) {
+            if (versions[replicaIndex - 1] == InternalPartition.SYNC_WAITING) {
+                biggestLostReplicaIndex = replicaIndex;
+            }
+        }
+
+        return biggestLostReplicaIndex;
+    }
+
+    private void overwriteLostReplicaVersionsWithFirstAvailableVersion(final long[] versions, final int lostReplicaIndex) {
+        final long forcedVersion = lostReplicaIndex < versions.length ? versions[lostReplicaIndex] : 0;
+        for (int replicaIndex = lostReplicaIndex; replicaIndex > 0; replicaIndex--) {
+            versions[replicaIndex - 1] = forcedVersion;
+        }
+    }
+
+    private void resetSyncWaitingVersions(long[] versions) {
+        for (int replicaIndex = 1; replicaIndex <= versions.length; replicaIndex++) {
+            if (versions[replicaIndex - 1] == InternalPartition.SYNC_WAITING) {
+                versions[replicaIndex - 1] = 0;
+            }
+        }
+    }
+
+    private void sendPartitionLostEvent(int partitionId, int lostReplicaIndex) {
+        final InternalPartitionLostEvent partitionLostEvent =
+                new InternalPartitionLostEvent(partitionId, lostReplicaIndex, getNodeEngine().getThisAddress());
+        NodeEngineImpl nodeEngine = (NodeEngineImpl) getNodeEngine();
+        for (PartitionAwareService service : nodeEngine.getServices(PartitionAwareService.class)) {
+            try {
+                service.onPartitionLost(partitionLostEvent);
```

Comment:
`PartitionAwareService`s are called on partition thread here. Is this intentional? Do you actually need to block partition thread? Or can this be offloaded to another thread?